docs/development/design/build-images.rst

```diff
@@ -48,131 +45,256 @@ This produced issues to people pinning their images to any of these names becaus
-   However, at this point it requires extra work to change our build pipeline.
-   We decided to install OS packages from the application itself for now using
-   Docker API to call ``docker exec`` as ``root`` user.
+* ``build-default``
 
-   This reduces the amount of work required but also allows us to add this feature
-   to our current existing images (they require a rebuild to add ``nodenv`` and ``rustup``)
+  * python 3.9.6: 2m21.331s
+  * mambaforge 4.10.1: 0m26.291s
+  * miniconda3 4.7.12: 0m9.955s
+  * nodejs 14.17.5: 0m5.603s
+  * rust 1.54.0: 0m13.587s
+  * golang 1.17: 1m30.428s
 
+* ``build-large``
+
+  * python 3.9.6: 2m33.688s
+  * mambaforge 4.10.1: 0m28.781s
+  * miniconda3 4.7.12: 0m10.551s
+  * nodejs 14.17.5: 0m6.136s
+  * rust 1.54.0: 0m14.716s
+  * golang 1.17: 1m36.470s
+
+Note that the only one that required compilation was Python.
+All the others, spent 100% of its time downloading the binary.
+These download times are *way better from EU* with a home internet connection.
+
+In the worst scenario: "none of the specified language version has a pre-built image",
+the build will require ~5 minutes to install all the language requirements.
+By providing *only* pre-built images with the Python version (that's the most time consuming),
+build times will only require ~2 minutes to install the others.
+However, requiring one version of each language is not a common case.
+
+Cache language binaries on S3
+-----------------------------
+
+``asdf`` scripts can be altered to download the ``.tar.gz`` dist files from a different mirror than the official one.
+Read the Docs can make usage of this to create a mirror hosted locally on S3 to get faster download speeds.
+This will make a good improvement for languages that offer binaries: ``nodejs``, ``rust`` and ``go``:
+
+* ``nodejs`` uses ``NODEJS_ORG_MIRROR``: https://github.com/asdf-vm/asdf-nodejs/blob/master/lib/utils.sh#L5
+* ``rust`` uses ``RUSTUP_UPDATE_ROOT``: https://github.com/rust-lang/rustup/blob/master/rustup-init.sh#L23
+* ``go`` has the URL hardcoded: https://github.com/kennyp/asdf-golang/blob/master/bin/download#L54
+
+However, currently Python does not offer binaries and a different solution is needed.
+Python versions can be pre-compiled once and expose the output on the S3 for the builders to download and extract in the correct PATH.
+
+.. tip::
+
+   Since we are building a special cache for pre-compiled Python versions,
+   we could use the same method for all the other languages instead of creating a full mirror (many Gigabyes)
+   This simple `bash script`_ download the language sources, compiles it and upload it to S3 without requiring a mirror.
+   Note that it works in the same way for all the languages, not just for Python.
+
+
+
+Questions
+---------
+
+What Python versions will be pre-compiled and cached?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+At start only a small subset of Python version will be pre-compiled:
+
+* 2.7.x
```

Comment:
I don't think we can drop support for 2.7. Unfortunately, _the world_ still depends on it.

I'm fine dropping other old 3.x versions, because people can use a newer one, but I don't think users depending on 2.7 are able to migrate to 3.x easily.

We could grab some data from our users and check how many of them are using 2.7, but still --I'm not sure it's a good idea to drop it.

As long as we can compile it on our Ubuntu images via `asdf`, it should be fine to support it 😄